### ISSUE TITLE
Analyzer missing hook to filter classes using Agent includes/excludes settings

### ISSUE THREAD

**Issue by systemsplanet** [jacoco/jacoco#930]
org.jacoco.core.analysis.Analyzer should include a way to skip (short circuit) analyzing classes based on 
the values used by the JaCoCo agent's **includes** and **excludes** settings.
Otherwise Reports coverage is incorrectly calculated when directories containing Jars are analyzed (unless you manually explode the jars and exclude classes yourself). I've solved this issued using the change below:

```java
private void analyzeClass(final byte[] source) {
        final long classId = CRC64.classId(source);
        final ClassReader reader = InstrSupport.classReaderFor(source);
        final String className = reader.getClassName();
        if (!filter.includeInReport(className)) return;  // <<<<<< SHORT CIRCUIT
        if ((reader.getAccess() & Opcodes.ACC_MODULE) != 0) {
            return;
        }
        if ((reader.getAccess() & Opcodes.ACC_SYNTHETIC) != 0) {
            return;
        }
        final ClassVisitor visitor = createAnalyzingVisitor(classId,className);
        reader.accept(visitor, 0);
    }
```

Note this Filter implementation only handles includes and excludes. 
It does not account for options.getExclClassloader(),options.getInclBootstrapClasses(), or options.getInclNoLocationClasses().

```java
import java.security.CodeSource;
import java.security.ProtectionDomain;
import org.jacoco.agent.rt.internal_035b120.CoverageTransformer;
import org.jacoco.core.runtime.WildcardMatcher;

public class Filter {
    private static final String AGENT_PREFIX;
    static {
        final String name = CoverageTransformer.class.getName();
        AGENT_PREFIX = toVMName(name.substring(0, name.lastIndexOf('.')));
    }
    private final WildcardMatcher includes;
    private final WildcardMatcher excludes;

    public Filter (String includesStr, String excludesStr) {
        includes = new WildcardMatcher(toVMName(includesStr));
        excludes = new WildcardMatcher(toVMName(excludesStr));
    }

    public boolean includeInReport( final String className) {
        boolean include = !className.startsWith(AGENT_PREFIX) && includes.matches(className) && !excludes.matches(className);
        return include;
    }
    
    private static String toVMName(final String srcName) {
        return srcName.replace('.', '/');
    }
}
```

**Comment by marchof** [comment]
The `Analyzer`is used during report generation. It is typically used in a different stage of the build and in a different JVM that the JaCoCo agent. Therefore the Analyzer has no access to the agent's `includes` and `excludes` settings.

> Otherwise Reports coverage is incorrectly calculated when directories containing Jars are analyzed (unless you manually explode the jars and exclude classes yourself).

If I understand your issue correctly this is the core of the request. Actually the way how class files are provided to the `Analyzer` highly depends on the integration  (Ant, Maven, CLI, ...). Maybe you can describe your setup a bit more in detail so we can get an understanding of your use case: What integration are you using? How does your setup look like? What classes are you going to filter (why?).

**Comment by systemsplanet** [comment]
I'm using the API, not Ant, not the maven plugin.

I've fixed the issue. I understand completely how it works. The agent uses
include/exclude patterns to instrument code, while the Report API has to
reproduce the exact same filtering except based on the file system.

If your report is for a spring boot war that includes a hundred jars, you
have to  explode the war, explode the jars into directories and figure out
where the classes are and filter paths using file system paths.

This is all unnecessary, if you just make the change I made. Then you can
run using the exact same includes/excludes strings as the Agent.

There is an outstanding Jenkins bug related to this issue.

Bottom line is you have two different mechanisms for selecting Classes and
only one of them works well, which is the Agent. The Report class selection
is a painful exercise left to the user.
